File: Vtt2LrcApp/app/src/main/java/com/qilong/vtt2lrc/MainActivity.java
package com.qilong.vtt2lrc;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.documentfile.provider.DocumentFile;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MainActivity extends Activity {
    private static final int REQ_PICK_FOLDER = 1001;
    private TextView logText;
    private int converted = 0;
    private int skipped = 0;
    private int failed = 0;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTitle("VTT转LRC");

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(dp(18), dp(18), dp(18), dp(18));
        root.setGravity(Gravity.CENTER_HORIZONTAL);

        TextView title = new TextView(this);
        title.setText("VTT转LRC");
        title.setTextSize(26);
        title.setGravity(Gravity.CENTER);
        title.setPadding(0, dp(8), 0, dp(8));
        root.addView(title, new LinearLayout.LayoutParams(-1, -2));

        TextView desc = new TextView(this);
        desc.setText("选择包含 .vtt 的文件夹，App会递归转换所有字幕，并在原目录生成同名 .lrc。输出为 UTF-8 BOM + LF 换行，适合网易云识别滚动歌词。");
        desc.setTextSize(15);
        desc.setPadding(0, 0, 0, dp(14));
        root.addView(desc, new LinearLayout.LayoutParams(-1, -2));

        Button pick = new Button(this);
        pick.setText("选择文件夹并转换");
        pick.setAllCaps(false);
        pick.setOnClickListener(v -> openFolderPicker());
        root.addView(pick, new LinearLayout.LayoutParams(-1, dp(52)));

        logText = new TextView(this);
        logText.setText("等待选择文件夹…");
        logText.setTextSize(14);
        logText.setPadding(0, dp(16), 0, 0);

        ScrollView scrollView = new ScrollView(this);
        scrollView.addView(logText);
        root.addView(scrollView, new LinearLayout.LayoutParams(-1, 0, 1));

        setContentView(root);
    }

    private void openFolderPicker() {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION
                | Intent.FLAG_GRANT_WRITE_URI_PERMISSION
                | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION
                | Intent.FLAG_GRANT_PREFIX_URI_PERMISSION);
        startActivityForResult(intent, REQ_PICK_FOLDER);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQ_PICK_FOLDER && resultCode == RESULT_OK && data != null && data.getData() != null) {
            Uri uri = data.getData();
            int flags = data.getFlags() & (Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
            try {
                getContentResolver().takePersistableUriPermission(uri, flags);
            } catch (Exception ignored) {
            }
            DocumentFile folder = DocumentFile.fromTreeUri(this, uri);
            converted = 0;
            skipped = 0;
            failed = 0;
            logText.setText("开始转换…\n");
            if (folder == null || !folder.isDirectory()) {
                appendLog("选择的不是有效文件夹。");
                return;
            }
            processFolder(folder);
            appendLog(String.format(Locale.CHINA, "\n完成：成功 %d，跳过 %d，失败 %d", converted, skipped, failed));
        }
    }

    private void processFolder(DocumentFile folder) {
        DocumentFile[] files = folder.listFiles();
        for (DocumentFile file : files) {
            if (file.isDirectory()) {
                processFolder(file);
            } else if (file.isFile() && file.getName() != null && file.getName().toLowerCase(Locale.ROOT).endsWith(".vtt")) {
                convertOne(file, folder);
            }
        }
    }

    private void convertOne(DocumentFile vttFile, DocumentFile parentFolder) {
        String fileName = vttFile.getName();
        if (fileName == null) return;
        try {
            String title = stripVttSuffix(fileName);
            String lrcName = title + ".lrc";
            String vttContent = readText(vttFile);
            String cleaned = cleanVtt(vttContent);
            String lrcBody = vttToLrc(cleaned);

            if (!Pattern.compile("\\[\\d+:\\d{2}\\.\\d{2}]").matcher(lrcBody).find()) {
                skipped++;
                appendLog("跳过：" + fileName + "，无有效时间轴");
                return;
            }

            String finalContent = "[ti:" + title + "]\n[ar:未知歌手]\n" + lrcBody + "\n";
            DocumentFile old = parentFolder.findFile(lrcName);
            if (old != null) old.delete();
            DocumentFile lrcFile = parentFolder.createFile("text/plain", lrcName);
            if (lrcFile == null) throw new IllegalStateException("无法创建 " + lrcName);

            try (OutputStream out = getContentResolver().openOutputStream(lrcFile.getUri(), "wt")) {
                if (out == null) throw new IllegalStateException("无法写入 " + lrcName);
                out.write(new byte[]{(byte) 0xEF, (byte) 0xBB, (byte) 0xBF});
                out.write(finalContent.replace("\r\n", "\n").replace("\r", "\n").getBytes(StandardCharsets.UTF_8));
            }
            converted++;
            appendLog("完成：" + fileName + " → " + lrcName);
        } catch (Exception e) {
            failed++;
            appendLog("失败：" + fileName + "，" + e.getMessage());
        }
    }

    private String readText(DocumentFile file) throws Exception {
        StringBuilder sb = new StringBuilder();
        try (InputStream in = getContentResolver().openInputStream(file.getUri())) {
            if (in == null) throw new IllegalStateException("无法读取文件");
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line).append('\n');
                }
            }
        }
        return sb.toString();
    }

    private String cleanVtt(String s) {
        String cleaned = s.replaceFirst("^\\uFEFF", "");
        cleaned = cleaned.replaceFirst("(?s)^WEBVTT.*?\\n\\s*\\n", "");
        cleaned = cleaned.replaceAll("(?m)^\\s*#.*?\\n", "");
        return cleaned;
    }

    private String vttToLrc(String vttContent) {
        StringBuilder lrc = new StringBuilder();
        Pattern pattern = Pattern.compile(
                "(\\d{2}:\\d{2}:\\d{2}\\.\\d{3}) --> (\\d{2}:\\d{2}:\\d{2}\\.\\d{3})\\s*(.*?)(?=\\n*\\d{2}:\\d{2}:\\d{2}\\.\\d{3} --> |$)",
                Pattern.DOTALL | Pattern.MULTILINE
        );
        Matcher matcher = pattern.matcher(vttContent);
        while (matcher.find()) {
            String startTime = matcher.group(1);
            String lyric = matcher.group(3);
            String lrcStart = convertTime(startTime);
            String lyricClean = cleanLyric(lyric);
            if (!lyricClean.isEmpty()) {
                lrc.append(lrcStart).append(' ').append(lyricClean).append('\n');
            }
        }
        return lrc.toString().trim();
    }

    private String cleanLyric(String lyric) {
        String[] lines = lyric.split("\\R");
        StringBuilder sb = new StringBuilder();
        for (String line : lines) {
            String t = line.trim();
            if (!t.isEmpty()) {
                if (sb.length() > 0) sb.append('\n');
                sb.append(t);
            }
        }
        return sb.toString();
    }

    private String convertTime(String vttTime) {
        String[] parts = vttTime.split(":");
        int hours = Integer.parseInt(parts[0]);
        int minutes = Integer.parseInt(parts[1]);
        String[] secMs = parts[2].split("\\.");
        int totalMinutes = hours * 60 + minutes;
        String sec = leftPad2(secMs[0]);
        String ms = (secMs[1] + "000").substring(0, 2);
        return "[" + totalMinutes + ":" + sec + "." + ms + "]";
    }

    private String leftPad2(String s) {
        return s.length() >= 2 ? s : "0" + s;
    }

    private String stripVttSuffix(String fileName) {
        if (fileName.toLowerCase(Locale.ROOT).endsWith(".vtt")) {
            return fileName.substring(0, fileName.length() - 4);
        }
        return fileName;
    }

    private void appendLog(String msg) {
        runOnUiThread(() -> logText.append(msg + "\n"));
    }

    private int dp(int value) {
        return (int) (value * getResources().getDisplayMetrics().density + 0.5f);
    }
}

File: app/src/main/java/com/example/vtt2lrc/MainActivity.kt
package com.example.vtt2lrc

import android.app.Activity
import android.content.Intent
import android.graphics.Typeface
import android.net.Uri
import android.os.Bundle
import android.view.ViewGroup
import android.widget.Button
import android.widget.CheckBox
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import androidx.activity.ComponentActivity
import androidx.activity.result.contract.ActivityResultContracts
import androidx.documentfile.provider.DocumentFile
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.regex.Pattern

class MainActivity : ComponentActivity() {

    private lateinit var overwriteBox: CheckBox
    private lateinit var recursiveBox: CheckBox
    private lateinit var logView: TextView

    private val folderPicker = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode != Activity.RESULT_OK) {
            appendLog("Selection cancelled.")
            return@registerForActivityResult
        }

        val uri = result.data?.data
        if (uri == null) {
            appendLog("No folder URI returned.")
            return@registerForActivityResult
        }

        persistTreePermission(uri, result.data?.flags ?: 0)
        val folder = DocumentFile.fromTreeUri(this, uri)
        if (folder == null || !folder.isDirectory) {
            appendLog("Selected item is not a valid folder.")
            return@registerForActivityResult
        }

        clearLog()
        appendLog("Start: ${folder.name ?: "selected folder"}")
        val stats = ProcessStats()
        processFolder(folder, stats)
        appendLog("")
        appendLog("Done: found ${stats.found} VTT, wrote ${stats.written} LRC, skipped ${stats.skipped}, failed ${stats.failed}.")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        buildUi()
    }

    private fun buildUi() {
        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(18), dp(18), dp(18), dp(18))
            setBackgroundColor(0xFFFFFFFF.toInt())
        }

        val title = TextView(this).apply {
            text = "VTT to LRC"
            textSize = 24f
            setTypeface(Typeface.DEFAULT_BOLD)
            setTextColor(0xFF111111.toInt())
        }

        val summary = TextView(this).apply {
            text = "Pick a folder, find .vtt files, and write same-name .lrc files beside them. Output is UTF-8 with BOM and LF line endings."
            textSize = 14f
            setTextColor(0xFF555555.toInt())
            setPadding(0, dp(8), 0, dp(12))
        }

        val pickButton = Button(this).apply {
            text = "Choose Folder"
            setOnClickListener { openFolderPicker() }
        }

        overwriteBox = CheckBox(this).apply {
            text = "Overwrite existing LRC"
            isChecked = true
        }

        recursiveBox = CheckBox(this).apply {
            text = "Include subfolders"
            isChecked = true
        }

        val logTitle = TextView(this).apply {
            text = "Log"
            textSize = 17f
            setTypeface(Typeface.DEFAULT_BOLD)
            setPadding(0, dp(14), 0, dp(8))
        }

        logView = TextView(this).apply {
            text = "Choose a folder that contains .vtt files."
            textSize = 14f
            setTextColor(0xFF222222.toInt())
            setPadding(dp(12), dp(12), dp(12), dp(12))
            setBackgroundColor(0xFFF5F5F5.toInt())
        }

        val scroll = ScrollView(this).apply {
            addView(
                logView,
                ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
            )
        }

        root.addView(title)
        root.addView(summary)
        root.addView(pickButton, LinearLayout.LayoutParams.MATCH_PARENT, dp(48))
        root.addView(overwriteBox)
        root.addView(recursiveBox)
        root.addView(logTitle)
        root.addView(scroll, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f))
        setContentView(root)
    }

    private fun openFolderPicker() {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE).apply {
            addFlags(
                Intent.FLAG_GRANT_READ_URI_PERMISSION or
                    Intent.FLAG_GRANT_WRITE_URI_PERMISSION or
                    Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION or
                    Intent.FLAG_GRANT_PREFIX_URI_PERMISSION
            )
        }
        folderPicker.launch(intent)
    }

    private fun persistTreePermission(uri: Uri, resultFlags: Int) {
        val requestedFlags = Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
        val takeFlags = resultFlags and requestedFlags
        try {
            contentResolver.takePersistableUriPermission(uri, if (takeFlags == 0) requestedFlags else takeFlags)
        } catch (e: SecurityException) {
            appendLog("Warning: could not persist folder permission: ${e.message}")
        }
    }

    private fun processFolder(folder: DocumentFile, stats: ProcessStats) {
        folder.listFiles().forEach { child ->
            when {
                child.isDirectory && recursiveBox.isChecked -> processFolder(child, stats)
                child.isFile && child.name?.lowercase()?.endsWith(".vtt") == true -> {
                    stats.found++
                    convertFile(child, folder, stats)
                }
            }
        }
    }

    private fun convertFile(vttFile: DocumentFile, parentFolder: DocumentFile, stats: ProcessStats) {
        val vttName = vttFile.name ?: return
        val lrcName = vttName.replace(Regex("(?i)\\.vtt$"), ".lrc")

        try {
            val lrcContent = buildLrcContent(vttName, readText(vttFile.uri))
            if (lrcContent == null) {
                stats.skipped++
                appendLog("Skip: $vttName has no valid timestamp.")
                return
            }

            val existing = parentFolder.findFile(lrcName)
            if (existing != null) {
                if (!overwriteBox.isChecked) {
                    stats.skipped++
                    appendLog("Skip: $lrcName already exists.")
                    return
                }
                if (!existing.delete()) {
                    stats.failed++
                    appendLog("Failed: cannot overwrite $lrcName.")
                    return
                }
            }

            val lrcFile = parentFolder.createFile("application/octet-stream", lrcName)
            if (lrcFile == null) {
                stats.failed++
                appendLog("Failed: cannot create $lrcName.")
                return
            }

            writeUtf8BomLf(lrcFile.uri, lrcContent)
            stats.written++
            appendLog("OK: $vttName -> $lrcName")
        } catch (e: Exception) {
            stats.failed++
            appendLog("Error: $vttName, ${e.message ?: e.javaClass.simpleName}")
        }
    }

    private fun readText(uri: Uri): String {
        contentResolver.openInputStream(uri)?.use { input ->
            return BufferedReader(InputStreamReader(input, Charsets.UTF_8)).readText()
        }
        throw IllegalStateException("Cannot read file")
    }

    private fun writeUtf8BomLf(uri: Uri, text: String) {
        val lfText = text.replace("\r\n", "\n").replace("\r", "\n")
        contentResolver.openOutputStream(uri, "wt")?.use { output ->
            output.write(byteArrayOf(0xEF.toByte(), 0xBB.toByte(), 0xBF.toByte()))
            output.write(lfText.toByteArray(Charsets.UTF_8))
        } ?: throw IllegalStateException("Cannot write file")
    }

    private fun buildLrcContent(fileName: String, raw: String): String? {
        val normalized = raw
            .replace("\r\n", "\n")
            .replace("\r", "\n")
            .replace(Regex("^\\uFEFF"), "")
            .replace(Regex("^WEBVTT.*?\\n\\n", RegexOption.DOT_MATCHES_ALL), "")
            .replace(Regex("^\\s*#.*?\\n", RegexOption.MULTILINE), "")

        val body = vttToLrc(normalized)
        if (!Regex("\\[\\d+:\\d{2}\\.\\d{2}]").containsMatchIn(body)) {
            return null
        }

        val title = fileName.replace(Regex("(?i)\\.vtt$"), "")
        return "[ti:$title]\n[ar:\u672a\u77e5\u6b4c\u624b]\n$body"
    }

    private fun vttToLrc(vttContent: String): String {
        val lines = mutableListOf<String>()
        val pattern = Pattern.compile(
            "(\\d{2}:\\d{2}:\\d{2}\\.\\d{3}) --> (\\d{2}:\\d{2}:\\d{2}\\.\\d{3})\\s*(.*?)(?=\\n*\\d{2}:\\d{2}:\\d{2}\\.\\d{3} --> |$)",
            Pattern.DOTALL or Pattern.MULTILINE
        )
        val matcher = pattern.matcher(vttContent)
        var previousTime = "0:00.00"

        while (matcher.find()) {
            val startTime = matcher.group(1) ?: continue
            val lyric = matcher.group(3) ?: ""
            val lrcStart = convertTime(startTime)
            if (lrcStart < previousTime) {
                appendLog("Warning: timestamp order changed, $previousTime -> $lrcStart")
            }
            previousTime = lrcStart

            val lyricClean = lyric
                .lineSequence()
                .map { it.trim() }
                .filter { it.isNotEmpty() }
                .joinToString("\n")

            if (lyricClean.isNotEmpty()) {
                lines.add("$lrcStart $lyricClean")
            }
        }

        return lines.joinToString("\n")
    }

    private fun convertTime(vttTime: String): String {
        val parts = vttTime.split(":")
        val hours = parts[0].toInt()
        val minutes = parts[1].toInt()
        val secMs = parts[2].split(".")
        val totalMinutes = hours * 60 + minutes
        val sec = secMs[0].padStart(2, '0')
        val ms = secMs[1].padEnd(3, '0').substring(0, 2)
        return "[$totalMinutes:$sec.$ms]"
    }

    private fun appendLog(message: String) {
        logView.append(if (logView.text.isNullOrBlank()) message else "\n$message")
    }

    private fun clearLog() {
        logView.text = ""
    }

    private fun dp(value: Int): Int = (value * resources.displayMetrics.density).toInt()

    private data class ProcessStats(
        var found: Int = 0,
        var written: Int = 0,
        var skipped: Int = 0,
        var failed: Int = 0
    )
}